JDK-8262955: Unify os::fork_and_exec() across Posix platforms

tstuefe opened the pull request:
`os::fork_and_exec()` can be used from within the hotspot to start a child process. It is only called in fatal situations, in two cases:
a) to automatically start a debugger when ShowMessageBoxOnError is specified (uses *fork*())
b) to start a caller provided binary on OOM if -XX:OnOutOfMemoryError is specified (uses *vfork*())

The variants for AIX, Linux, Bsd are almost completely identical. So, this function can be unified under posix.

In addition to that, this patch does a number of small changes:

1) Before, whether we would vfork() only on Linux and only for case (b). I changed this to always use vfork unconditionally, on all platforms, because: 
  - even though vfork() can be unsafe, the way we use it - calling  vfork()->exec()->_exit() with no intermediate steps - is safe.
  - Using vfork is good for OOM situations on all platforms, not just Linux, and also for starting the debugger in non-OOM cases. Keep in mind that we do this only for cases where the parent VM is about to die, so even if it were unsafe, the damage would be limited.
2) I added a comment to the function to not use it outside of fatal error situations.
3) I added a posix wrapper for getting the environ pointer, to hide MacOS specifics, and used it in two places to unify that coding.
4) consistently used global scope :: for posix APIs.

Note that if we wanted to make os::fork_and_exec() a first class function, always safe to use, we should modify it to at least not leak any parent process file descriptors. Possibly safest would be to completely rewrite this function and use posix_spawn(). posix_spawn() we use in Runtime.exec() by default since JDK 13 (1). But as long as this is spawned by only dying VMs I think this function is fine.

----

Tests: GAs, manual tests using -XX:ShowMessageBoxOnError

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8262955](https://bugs.openjdk.java.net/browse/JDK-8262955): Unify os::fork_and_exec() across Posix platforms


### Reviewers
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Harold Seigel](https://openjdk.java.net/census#hseigel) (@hseigel - **Reviewer**)


### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/2810/head:pull/2810`
`$ git checkout pull/2810`

tstuefe commented:
> _Mailing list message from [David Holmes](mailto:david.holmes@oracle.com) on [hotspot-dev](mailto:hotspot-dev@openjdk.java.net):_
> 
> Hi Thomas,
> 
> On 5/03/2021 4:22 pm, Thomas Stuefe wrote:
> 
> > `os::fork_and_exec()` can be used from within the hotspot to start a child process. It is only called in fatal situations, in two cases:
> > a) to automatically start a debugger when ShowMessageBoxOnError is specified (uses *fork*())
> > b) to start a caller provided binary on OOM if -XX:OnOutOfMemoryError is specified (uses *vfork*())
> > The variants for AIX, Linux, Bsd are almost completely identical. So, this function can be unified under posix.
> > In addition to that, this patch does a number of small changes:
> > 1) Before, whether we would vfork() only on Linux and only for case (b). I changed this to always use vfork unconditionally, on all platforms, because:
> > - even though vfork() can be unsafe, the way we use it - calling  vfork()->exec()->_exit() with no intermediate steps - is safe.
> 
> Is it? The reason we use fork() for the error/crash case is because it
> can get launched from a signal handling context and vfork is not
> async-signal-safe.
> 
> There is some commentary in:
> 
> https://bugs.openjdk.java.net/browse/JDK-8027434
> 
> Cheers,
> David

Hi David,

My estimate of vfork being safe here comes from experience. At SAP, at one time we replaced the jdk's Runtime.exec() implementation completely with our own; it lived in hotspot and was used by both the jdk and the hotspot, in and out of signal contexts. Our implementation mainly used vfork(), with a lot of safe guards of course (mainly a forkhelper binary, similarly to what Runtime.exec() does today).

I cannot completely exclude the possibility of problems here, but calling vfork()->exec()->_exit() is as safe as it gets. Seeing that we only use it for starting a debugger in case the original VM crashed, I think the possibility of problems is remote.

OTOH when starting a debugger spawned from a fat process, you may run into the same problems as with https://bugs.openjdk.java.net/browse/JDK-8027434. You don't want the machine to go into swap when trying to start the debugger for you.

I won't fight you on this though if you insist; mainly what I disliked was the introduction of Posix terminology in Windows code ("fork_and_exec" "use_vfork") and that can be straightened out by a separate layer (e,g. os::start_child_process(bool from_error_handler) -> os::Posix::os_fork_and_exec(can_use_vfork)).

Cheers, Thomas

hseigel commented:
Hi Thomas,
Can the #includes of <sys/wait.h> be removed from the os_aix.cpp, os_bsd.cpp, and os_linux.cpp files?
Thanks, Harold

tstuefe commented:
> Hi Thomas,
> Can the #includes of <sys/wait.h> be removed from the os_aix.cpp, os_bsd.cpp, and os_linux.cpp files?
> Thanks, Harold

Yes, good catch, I'll remove them.

tstuefe commented:
Hi,

I added the following changes:
- took care of Harold's request
- modified the coding to only use vfork when we are assured outside of signal context. For that, I added a marker into Thread which marks if the Thread is inside hotspot signal handling. If we want, we can use this in the future for better, signal handler aware tests (eg assert that we don't use malloc in tests).

I also added POSIX macros to macros.hpp since this is overdue. We can use those to replace constructs where we use #ifndef WINDOWS to guard posix coding. Seems cleaner that way to me.

Thanks, Thomas

dholmes-ora commented on src/hotspot/os/posix/os_posix.cpp:
```diff
@@ -1765,3 +1771,81 @@ int os::PlatformMonitor::wait(jlong millis) {
+  // But since vfork is not async-safe we only use it where we are sure to be outside
+  //  of signal handling.
+  const Thread* const t = Thread::current_or_null_safe();
+  const bool use_vfork = t != NULL && t->is_in_signal_handler() == false;
```
Nit: !t->is_in_signal_handler()

dholmes-ora commented on src/hotspot/os/posix/signals_posix.cpp:
```diff
@@ -565,6 +565,22 @@ class ErrnoPreserver: public StackObj {
+public:
+  InSignalHandlerMark() :
+    _thread(Thread::current_or_null_safe()) {
+    if (_thread) {
```
Style Nit: no implicit booleans - use `_thread != NULL`

dholmes-ora commented on src/hotspot/os/posix/signals_posix.cpp:
```diff
@@ -581,6 +597,8 @@ int JVM_HANDLE_XXX_SIGNAL(int sig, siginfo_t* info,
 {
   assert(info != NULL && ucVoid != NULL, "sanity");
 
+  InSignalHandlerMark ishm;
```
Is this going to behave as expected if we actually crash during signal processing and so re-enter this routine with a signal mark already active?

dholmes-ora commented on src/hotspot/share/utilities/macros.hpp:
```diff
@@ -431,6 +431,17 @@
+#define POSIX
+#endif
+#define POSIX_ONLY(code) code
+#define NOT_POSIX(code)
```
Not sure this makes sense. At the moment only Windows is not-POSIX. If in the future we had another non-POSIX platform then I find it very unlikely that it would use the same code as Windows. So I would not like to see NOT_POSIX used incorrectly where we really only mean WINDOWS.

tstuefe replied:
My intention was exactly that, using it to protect *Posix* code from compiling on non-Posix platforms. Using !Windows there is just as wrong and not as descriptive. Most !Windows cases in hotspot are not affected by this since they have nothing to do with Posix; but there are a few and there may be more. E.g. I am currently working on a patch to protect the program break on Unices. This means dealing with sbrk, and depending on which route I go a POSIX_ONLY macro would make perfect sense.

That POSIX is defined as !Windows is only incidental. If you prefer, I can list the posix operating systems here; I did not do that for brevity.

dholmes-ora commented on src/hotspot/share/runtime/thread.hpp:
```diff
@@ -822,6 +822,15 @@ class Thread: public ThreadShadow {
   static void SpinAcquire(volatile int * Lock, const char * Name);
   static void SpinRelease(volatile int * Lock);
+
+#ifdef POSIX
```
I hate seeing this in shared code. It really belongs in platform-dependent thread code - but osThread_posix doesn't exist. :(

But do we need this - can't the existing use_vfork_if_available parameter instead be renamed and interpreted as not_in_a_signal_handler?

tstuefe replied:
But would having this functionality not make sense? We always argue about what is and is not allowed during signal handling; here, we may get a mechanism to actually assert it and e.g. prevent using malloc() or logging or whatever VM functionality may creep in.

But I will revert my last commit completely and re-add the needs-vfork parameter. This was supposed to be mainly a cleanup; maybe it was wrong to mix in behavioral changes.

dholmes-ora reviewed (commented):
Hi Thomas,

I appreciate the reworking to avoid vfork for the signal case, but have some concerns with the mechanism and I'm not sure a new mechanism is really needed. See comments below.

Thanks,
David

tstuefe commented:
> _Mailing list message from [David Holmes](mailto:david.holmes@oracle.com) on [hotspot-dev](mailto:hotspot-dev@openjdk.java.net):_
> 
> On 8/03/2021 4:12 pm, Thomas Stuefe wrote:
> 
> > On Mon, 8 Mar 2021 05:37:37 GMT, David Holmes <dholmes at openjdk.org> wrote:
> > > > Thomas Stuefe has updated the pull request incrementally with one additional commit since the last revision:
> > > > tidy up
> > > 
> > > 
> > > src/hotspot/share/runtime/thread.hpp line 826:
> > > > 824:   static void SpinRelease(volatile int * Lock);
> > > > 825:
> > > > 826: #ifdef POSIX
> > > 
> > > 
> > > I hate seeing this in shared code. It really belongs in platform-dependent thread code - but osThread_posix doesn't exist. :(
> > > But do we need this - can't the existing use_vfork_if_available parameter instead be renamed and interpreted as not_in_a_signal_handler?
> > 
> > 
> > But would having this functionality not make sense? We always argue about what is and is not allowed during signal handling; here, we may get a mechanism to actually assert it and e.g. prevent using malloc() or logging or whatever VM functionality may creep in.
> 
> I agree having IsInSignalHandler may be generally useful, my issue was
> with how you actually did it - in the shared code - and so simpler
> solution for the currnet case would just be to reuse the existing para,eter.
> 
> > But I will revert my last commit completely and re-add the needs-vfork parameter. This was supposed to be mainly a cleanup; maybe it was wrong to mix in behavioral changes.
> 
> True. Do the refactoring then look at enhancements later. Sorry for
> making this harder.
> 

Oh no problem, that's what reviews are for. There will be occasion enough for behavioral change RFEs later.

Cheers, Thomas

tstuefe commented:
New version; restored the vfork control boolean parameter; the only behavioral change to before is that on AIX we always use vfork since this is fine for AIX, and we have no overcommit there so its more urgent in case of OOMs.

dholmes-ora commented on src/hotspot/os/windows/os_windows.cpp:
```diff
@@ -5512,7 +5514,7 @@ int os::PlatformMonitor::wait(jlong millis) {
 // Run the specified command in a separate process. Return its exit value,
 // or -1 on failure (e.g. can't create a new process).
-int os::fork_and_exec(char* cmd, bool use_vfork_if_available) {
+int os::fork_and_exec(const char* cmd, bool dummy /* ignored */) {
```
You could just have `bool _ignored`

dholmes-ora reviewed (approved):
Hi Thomas,

I am fine with this version of the refactoring.

Thanks,
David

hseigel reviewed (approved):
Changes look good.  Thanks for doing this.
Harold

tstuefe commented:
> Hi Thomas,
> 
> I am fine with this version of the refactoring.
> 
> Thanks,
> David

Thank you David!

tstuefe commented:
> Changes look good. Thanks for doing this.
> Harold

Thanks Harold!

tstuefe commented:
/integrate
